### ISSUE TITLE
Figure out repo structure

### ISSUE THREAD

**Issue by kelvich** [zenithdb/postgres#24]

**Comment by adsharma** [comment]
Here's the high level proposal:

zenith.src -> top level repo with a bunch of submodules. Each of them is a git module that can be separately cloned.

1. postgres - tracks vanilla postgres. Have a zenith branch with minimal changes we need to carry forward. Each change to have a plan to push upstream or move to an extension.
2. pageserver
3. walkeeper (trying to keep the directory name short, we can use safekeeper elsewhere if we prefer)
4. consensus
5. cli
6. Top level README which explains what the submodules are and how to build them.

Let me know if I missed anything or if anything needs to be renamed.

**Comment by kelvich** [comment]
Why zenith.src? Let's call top repo just zenith.

Right now 2-5 would share most of the code, so I would rather model them as different files in src/bin -- cargo allows one crate to have several binaries. Later we may split some of that into a separate crates, but we have to write it first to determine right separation boundary. That way everything would be buildable by one command, namely 'cargo build'.

About postgres: yep let's place it as a submodule into a vendor/ dir or something like that. Some integration tests would require minio as well, but since we are not changing minio, I would rather not include it as a submodule and leave to a test setup scripts to ensure minio presence.

**Comment by adsharma** [comment]
Because there are probably other directories containing web pages, rfcs etc. Some of it may be private.

src/bin - that seems like a postgres concept. How do you imagine that working if we have a submodule which has a minimal delta over upstream?

It should be perfectly ok for pageserver and walkeeper to link to consensus via  line such as:

```
consensus = { path="../consensus" }
```

**Comment by adsharma** [comment]
I don't feel strongly about zenith.src - ok with zenith as well. Just explaining why Carl might have created it that way.

**Comment by kelvich** [comment]
> src/bin - that seems like a postgres concept.
I was talking about rust part of things: https://doc.rust-lang.org/cargo/reference/cargo-targets.html#binaries

So cargo is able to build few binaries (pageserver, walkeeper, cli, etc) out of a single crate. We may later put that stuff into different crates and put common stuff to some library-crate. But right now is not clear what is common stuff and what is not, so I would postpone that to later.

**Comment by adsharma** [comment]
I would be ok with building even a single binary - like `storage-server`

But why one crate? Different crates that have independent unit tests and can be developed in parallel in different timezones sounds like a big plus to me.

**Comment by adsharma** [comment]
Just in case it wasn't clear, in the proposal each one of these is a separate git repo/submodule.

So I'll be able to:

```
git clone https://github.com/libzenith/consensus
cargo test

git clone https://github.com/libzenith/walkeeper
cargo test
```
if I wanted to. Perhaps there is a CI/CD pipeline to do just that. But it's more convenient to clone the zenith repo, do:

```
git submodule update --init --recursive
```

and get everything.

**Comment by kelvich** [comment]
Both walkeeper and pageserver share right now big amount of code (both dealing with WAL and network mainly). So some parts of the code should go also in "common" crate. I believe that is the right way at end but now I would prefer to just have everything in one crate.

We are not depending on postgres as a build dependency, so all our stuff can be built without submodules and in one repo:
```
git clone https://github.com/libzenith/zenith
cargo build
```
and set of binaries would appear in a `target` directory.

But to run integration tests you would need to fetch vendored postgres:
```sh
git submodule update --init --recursive # fetches postgres
make postgres # build postgres and installs it to ./tmp_install locally
cargo test # run tests of pg+walkeeper+pageserver
```
If at some point in time we will be able to work with ordinary postgres we may drop this submodule and test against local postgres.

**Comment by kelvich** [comment]
Also there are two slightly different issues here: separation of code by crates and by repos.

1) With crates I would prefer having one right now and switch to few in let's say $now + 2 month when it will be clear what is common stuff there and how to better separate code.

2) Putting different crate in different repos is another thing. If we do that it would mean that if I change something in some zenith-common crate I would need to create commit there, then commit changes to every repo that uses it, then commit submodule update to a top-level repo. It is questionable structure even with million lines code base, but for several thousands loc codebase it is really strange. Cargo have special concept called workspaces for managing several crate in one repo, so we better stick to that.

**Comment by adsharma** [comment]
I believe having a common crate and putting everything there is an anti
pattern.

If there is shared code, let's put it in a well defined crate with a name
and purpose and share it.

This is a public repo. So someone looking to join the company can look at
and form an opinion about the company.

**Comment by kelvich** [comment]
Okay. Let's do it multi-crate. So that way zenith repo would be cargo workspace with several crates and postgres submodule goes to vendor/postgres, I suppose.

**Comment by adsharma** [comment]
Sounds good. Let me put this together and send a pull request.

**Comment by adsharma** [comment]
https://github.com/libzenith/zenith/pull/1

**Comment by hlinnaka** [comment]
Done, for now
